mpi4py parallelization, pipeline executable, slurm script generation and debugging mode

MichaelMedford opened the pull request:
This branch has four major changes implemented. My apologies for so many different changes falling into the same pull request, but I unfortunately got too deep into the problem before I realized how intertwined the issues had become. Future pull requests should cover only **one** feature at a time, because reviewing code with multiple feature changes in it is quite difficult (as you will soon see).

**Parallelization using the mpi4py protocol**
The previous implementation of parallelization in `synthetic.calc_events` used the `multiprocessing` package and an instance of the `pool` class. This method was not guaranteed to work on the larger supercomputers where we want to run many parallel instances of PopSyCLE. All of these changes can be found in these lines in `calc_events`: 

https://github.com/jluastro/PopSyCLE/blob/ae8dc764301548e6127535fb7e04a318bdb426ed/popsycle/synthetic.py#L1188-L1229

The downside of this implementation is that it changes how scripts must be called in order to run `calc_events` in parallel. The user can no longer simply include calling `calc_events` with `n_proc` set to the number of cores that they wish to execute. Instead the user must place `calc_events` into a separate python file that is called from the command line using mpi4py. For example: `mpiexec -n4 python perform_calc_events.py` where `perform_calc_events.py` includes the `calc_events` command. See the executable outlined below for a cleaner way to perform this execution.

 **PopSyCLE Pipeline Executable: run.py**
There are four steps to completing a PopSyCLE run:
1. Running Galaxia
2. Running `synthetic.perform_pop_syn`
3. Running `synthetic.calc_events` (possibly in parallel)
4. Running `synthetic.refine_events`

In order to facilitate running PopSyCLE on supercomputers, this branch provides a command line executable that runs these steps broken into three stages. This name of the executable is `run.py`. 

The three stages are:
- Stage 1 : Galaxia, `synthetic.perform_pop_syn`
- Stage 2 : `synthetic.calc_events`
- Stage 3 : `synthetic.refine_events`

The purpose of breaking the pipeline into three stages is so that the user can run the second stage (containing `calc_events`) with multiple processors. Running `run.py` requires that the destination folder contains a `galactic_configuration_file` that is generated using the function `slurm.generate_galactic_configuration_file`. This script also requires a `popsycle_configuration_file`, an example of which is provided in the `popsycle/docs` folder.

**Slurm Script Generator: slurm.py**
Slurm is an open source, fault-tolerant, and highly scalable cluster management and job scheduling system for large and small Linux clusters. Multiple instances of the PopSyCLE pipeline can be executed in parallel if running on a system that both has PopSyCLE and a slurm scheduler installed. The `popsycle.slurm.generate_slurm_scripts` function will create and submit slurm batch scripts that execute the entire PopSyCLE pipeline. I have included a detailed explanation of this executable and how it works in the new README, which I suggest you read over below. A key addition is the requirement of a `slurm_configuration_file`, an example of which is provided in the `popsycle/docs` folder.

**Debugging Mode**
Development of additional features requires a mode for running all stages of the PopSyCLE pipeline in an identically reproducible way. Introducing: debugging mode. In order to create identical results, Galaxia needs to be run with a fixed random seed in its configuration file and `perform_pop_syn` must be run with `debug=True`. Both of these conditions are set when running `run.py` with the `--debug` flag

**Additional Reformatting**
I also took this opportunity to make changes to `synthetic.py` that brought the formatting more into alignment with PEP8 standards, including spacing and line limits. My apologies for how difficult this makes `synthetic.py` to review for changes relevant to the features in this branch.

jluastro commented:
A few questions:
1. How easy/hard is it to run parallel on our desktops in the new branch?
2. You have switched to using MPI instead of multiprocessing… but this requires an MPI installation? Not such a big deal; but we need to be careful about MPI versions, etc. (compiled code is always more of a pain).

MichaelMedford commented:
1. Installing mpi4py on a personal desktop is quite easy (https://mpi4py.readthedocs.io/en/stable/install.html). Then it would simply mean running `mpiexec -n 4 python run.py --output-root=base0 --popsycle-config-filename=popsycle_config.yaml --stage=2`. Stage 2 runs `calc_events`, which is the only function that should be run in parallel. Alternatively the user could put `calc_events` into a separate file and running it with mpiexec. The other stages could be run with `run.py`, or simply called as python functions like before.

2. Installing MPI will be a necessary step to install mpi4py but there are walkthroughs on how to do so (https://mpi4py.readthedocs.io/en/stable/appendix.html#building-mpi). I believe that mac computers install MPI upon installing xcode and other necessary pacakges.

jluastro commented:
On 1. above:
This is a large number of flags for runtime. Can't we put in defaults for easier parallel runs on single machines? Also, I thought that we previously didn't have to run using mpiexec to run in parallel, correct? Meaning, we didn't need to have a command line script. You could run parallel through a Jupyter notebook  for instance. We seem to have lost this functionality.

Also, will the code compile if mpi4py is not installed? It should! The nice thing about multiprocessing was that it didn't require additional packages. 

On #2, MPI installation is notoriously painful. It is fine if have this dependency now; but we should clearly state that MPI is a dependency (for running in parallel), etc. And we should make it so that users can install without MPI if they wish (and run single-threaded).

MichaelMedford commented:
1. I will make the following defaults to `run.py`, forcing the user to only specify the stage upon runtime.
--output-root=base0
--popsycle-config-filename=popsycle_config.yaml

2. This update has removed the functionality to run `calc_events` in parallel within a jupyter notebook. (As far as I could research, it is not possible to run mpi4py in a jupyter notebook in this way.) The switch was made to mpi4py because multiprocessing cannot spread parallelism across multiple nodes. From the NERSC documentation:

> Python's standard library provides a multiprocessing package that supports spawning of processes. This can be used to achieve some level of parallelism within a single compute node. It cannot be used to achieve parallelism across compute nodes. For that, users are referred to the discussion on mpi4py here.

I think we could take this on in two ways. First, we could provide both mpi4py and multiprocessing parallelism as options with a parameter that the user sets to decide which type of parallelism to use. Then within notebooks we could use the multiprocessing option. Second, we could say that we will never run a single PopSyCLE run across multiple nodes, and turn the `calc_events` parallelism back into multiprocessing. Launching different PopSyCLE runs onto different nodes can be achieved through the slurm script mechanism that is already written into this merge.

3. The current setup allows for running the code without MPI installed. The user would simply run all scripts with parallelFlag = False, and the mpi4py code will be skipped over and not even attempted to be imported. MPI does *not* need to be installed to use the code as written.

MichaelMedford commented:
> On 1. above:
> This is a large number of flags for runtime. Can't we put in defaults for easier parallel runs on single machines? Also, I thought that we previously didn't have to run using mpiexec to run in parallel, correct? Meaning, we didn't need to have a command line script. You could run parallel through a Jupyter notebook for instance. We seem to have lost this functionality.
> 
> Also, will the code compile if mpi4py is not installed? It should! The nice thing about multiprocessing was that it didn't require additional packages.
> 
> On #2, MPI installation is notoriously painful. It is fine if have this dependency now; but we should clearly state that MPI is a dependency (for running in parallel), etc. And we should make it so that users can install without MPI if they wish (and run single-threaded).

MichaelMedford commented:
Per a conversation with Jessica, this branch now has the following features:

1. Removed all presence of mpi4py, instead reverting back to the original multiprocessing parallelism
2. Replaced debugging processes (which altered some functions) to setting a fixed seed for random processes that is called before each of the major user-facing functions (perform_pop_syn, calc_events, and refine_events)
3. Removed all stages from slurm.py, and then merged slurm.py functions into synthetic.py. I did this because it was a file with only one function, and that seemed unnecessary.
4. Put into place functions for saving and reading in configuration files. Confirmed that all information from the configuration files is present in the output logs.
5. Made the overwrite flags work as advertised.

I ran the code and confirmed that the random seed setting

jluastro commented on popsycle/synthetic.py:
```diff
@@ -133,8 +147,8 @@ def write_galaxia_params(output_root,
          Seed Galaxia will use to generate objects. If not set, script will
-         generate a random number from 0 to 100. This may want to be set for
-         reproducibility.
+         generate a random number from 0 to 100. Setting this seed guarantees
```
Standard procedure is to grab the system time for the random seed.... I think you could still do this (check online; but maybe just call seed with None?).

MichaelMedford replied:
Galaxia requires a two digit number from 0 to 100. This number will now be drawn from the current time.
```
if seed is None:
        # Grab last two digits of the current Epochal time
        seed = int(str(time.time())[-2:])
```

jluastro commented on popsycle/synthetic.py:
```diff
@@ -214,9 +229,15 @@ def perform_pop_syn(ebf_file, output_root, iso_dir,
+        Default is False.
+
+    seed : int
+        If set to non-zero, removes all random sampling and forces identical
```
This comment isn't quite accurate. Suggest changing to:

If set to non-zero, then all random sampling will be seeded with the specified seed.

MichaelMedford replied:
Changed

jluastro commented on popsycle/synthetic.py:
```diff
@@ -299,27 +338,30 @@ def perform_pop_syn(ebf_file, output_root, iso_dir,
     # Define bin_edges_number, if not given in input.
     if bin_edges_number is None:
-        bin_edges_number = int(60 * 2 * radius) + 1   # st the widths are 1/2 arcmin
+        bin_edges_number = int(
```
minor thing... I am not a fan of this line break. Move the comment up or down; but leave the equation on the same line.

MichaelMedford replied:
fixed

jluastro commented on popsycle/synthetic.py:
```diff
@@ -560,56 +619,71 @@ def calc_current_initial_ratio(iso_dir, out_file='current_initial_stellar_mass_r
+
 
-def current_initial_ratio(logage, ratio_file, iso_dir):
+def current_initial_ratio(logage, ratio_file, iso_dir, debug=False):
```
I thought we were changing this to random_seed?

caseylam replied:
Related: lines 303 - 309? Is that debug or random seed?

MichaelMedford replied:
See comment below

jluastro commented on popsycle/synthetic.py:
```diff
@@ -676,9 +759,13 @@ def _make_comp_dict(iso_dir, log_age, currentClusterMass, star_dict, next_id,
-        Forces PyPopStar to fix the random seed to 42,
-        enforcing identical output.
+    debug : bool
+        If set to True, removes all random sampling and forces identical
```
I thought we were removing debug and using random_seed instead?

MichaelMedford replied:
See comment below

jluastro commented on popsycle/synthetic.py:
```diff
@@ -693,14 +780,17 @@ def _make_comp_dict(iso_dir, log_age, currentClusterMass, star_dict, next_id,
 
     # Calculate the initial cluster mass
-    massLimits = np.array([0.1, 0.5, 120]) # changed from 0.08 to 0.1 at start because MIST can't handle.
+    massLimits = np.array([0.1, 0.5,
```
minor thing: don't break in the code... just move the comment up or down.

MichaelMedford replied:
fixed

jluastro commented on popsycle/synthetic.py:
```diff
@@ -1361,34 +1508,38 @@ def _calc_event_cands_radius(bigpatch, timei, radius_cut):
-    b_t = bigpatch[col_idx['glat']] + timei * bigpatch[col_idx['mu_b']] * masyr_to_degday # deg
-    l_t = bigpatch[col_idx['glon']] + timei * (bigpatch[col_idx['mu_lcosb']] / np.cos(np.radians(bigpatch[col_idx['glat']]))) * masyr_to_degday # deg
-    
+    r_t = bigpatch[col_idx['rad']] + timei * bigpatch[
```
minor: I'm not a fan of these line breaks.

MichaelMedford replied:
fixed

jluastro commented on popsycle/synthetic.py:
```diff
@@ -1451,41 +1604,48 @@ def _calc_event_cands_thetaE(bigpatch, theta_E, u, theta_frac, lens_id, sorc_id,
-        mu_lcosb_rel = sorc_table[col_idx['mu_lcosb']] - lens_table[col_idx['mu_lcosb']] # mas/yr
-        mu_rel = np.sqrt(mu_b_rel**2 + mu_lcosb_rel**2) # mas/yr
+
+        mu_b_rel = sorc_table[col_idx['mu_b']] - lens_table[
```
minor: not a fan of this line break.

MichaelMedford replied:
fixed

jluastro commented on popsycle/synthetic.py:
```diff
@@ -1451,41 +1604,48 @@ def _calc_event_cands_thetaE(bigpatch, theta_E, u, theta_frac, lens_id, sorc_id,
+
+        mu_b_rel = sorc_table[col_idx['mu_b']] - lens_table[
+            col_idx['mu_b']]  # mas/yr
+        mu_lcosb_rel = sorc_table[col_idx['mu_lcosb']] - lens_table[
```
minor: not a fan of this line break.

MichaelMedford replied:
fixed

jluastro commented on popsycle/synthetic.py:
```diff
@@ -1451,41 +1604,48 @@ def _calc_event_cands_thetaE(bigpatch, theta_E, u, theta_frac, lens_id, sorc_id,
+
         # This is all the events for this l, b, time
-        event_lbt = np.vstack((lens_table, sorc_table, theta_E, u, mu_rel, t_event))
+        event_lbt = np.vstack(
```
minor: not a fan of this line break.

MichaelMedford replied:
fixed

jluastro commented on popsycle/synthetic.py:
```diff
@@ -1564,13 +1724,15 @@ def _calc_blends(bigpatch, c, event_lbt, blend_rad):
-        lid = np.array(event_lbt[col_idx['obj_id'], ii]) # lens star object id
-        sid = np.array(event_lbt[col_idx['obj_id'] + len(col_idx), ii]) # source star object id
-        
+        nid = bigpatch[
```
minor: not a fan of this line break.

MichaelMedford replied:
fixed

jluastro commented on popsycle/synthetic.py:
```diff
@@ -1942,21 +2154,24 @@ def refine_events(input_root, filter_name, red_law,
     line11 = output_file + ' : refined events'
 
-    with open(input_root + '_refined_events_' + filter_name + '_' + red_law + '.log','w') as out:
+    with open(
```
minor: not a fan of this line break.

MichaelMedford replied:
fixed

jluastro commented on popsycle/synthetic.py:
```diff
@@ -2732,9 +2991,11 @@ def calc_ext(E, f):
+
 
-def sample_spherical(npoints, speed, ndim = 3):
+def sample_spherical(npoints, speed,
```
minor: not a fan of this line break.

MichaelMedford replied:
fixed

caseylam commented on popsycle/synthetic.py:
```diff
@@ -180,8 +194,9 @@ def write_galaxia_params(output_root,
-                    set_random_seed = False):
+                    bin_edges_number=None, BH_kick_speed=100,
+                    NS_kick_speed=350,
+                    overwrite=False, seed=None):
```
This might cause problems when merging w/ birth_kicks branch, but I think we already know/expect that.

MichaelMedford replied:
Yup, I'll take care of fixing that upon merging

caseylam commented on popsycle/synthetic.py:
```diff
@@ -768,9 +864,10 @@ def _make_comp_dict(iso_dir, log_age, currentClusterMass, star_dict, next_id,
-            comp_dict['px'], comp_dict['py'], comp_dict['pz'] = galactic_to_heliocentric(comp_dict['rad'],
-                                                                                         comp_dict['glat'],
-                                                                                         comp_dict['glon'])
+            comp_dict['px'], comp_dict['py'], comp_dict[
```
what happened here? can you do this linebreak?

MichaelMedford replied:
fixed

caseylam commented on popsycle/synthetic.py:
```diff
@@ -793,14 +890,15 @@ def _make_comp_dict(iso_dir, log_age, currentClusterMass, star_dict, next_id,
-                                                                                         comp_dict['rad'],
-                                                                                         comp_dict['glat'],
-                                                                                         comp_dict['glon'])
+            comp_dict['vr'], comp_dict['mu_b'], comp_dict[
```
same as above

MichaelMedford replied:
fixed

caseylam commented on popsycle/synthetic.py:
```diff
@@ -834,37 +932,50 @@ def _make_comp_dict(iso_dir, log_age, currentClusterMass, star_dict, next_id,
-                comp_dict['ubv_v'][lum_WD_idx] = comp_table['m_ubv_V'][lum_WD_idx].data
-                comp_dict['ubv_h'][lum_WD_idx] = comp_table['m_ukirt_H'][lum_WD_idx].data
-                
+                comp_dict['ubv_i'][lum_WD_idx] = comp_table['m_ubv_I'][
```
same as above

MichaelMedford replied:
fixed

MichaelMedford commented:
@jluastro @caseylam 
Here is the issue with the seed and debug flag. I have written PopSyCLE to be able to accept a user generated seed as an argument to `perform_pop_syn` and `calc_events` that seeds the random generators at the top of each function (`refine_events` has no random processes and thus does not have seed as an argument). However PyPopStar does *not* let the user provide a seed, but instead lets the user set a flag called `set_random_seed` to True or False. If set to True, PyPopStar forces the seed to be 42. So we have two options:

1. Change PyPopStar to accept user generated seeds and potentially break the code of people that are already using the `set_random_seed` flag with PyPopStar
2. At some point in the PopSyCLE pipeline, convert the seed that the user has provided into a debug boolean, with something like:
```
if seed is None:
    debug = False
else:
    debug = True
```
followed by
```
cluster = synthetic.ResolvedCluster(my_iso, trunc_kroupa, cluster_mass,
                                            set_random_seed=debug)
```
We could create this debug boolean at the lowest level possible, right before ResolvedCluster is called. Or we could do this at the place in the pipeline where the uniqueness of a seed doesn't make a difference. I choose the second option in the current commit.

For example, `calc_current_initial_ratio` currently has a boolean argument called debug and passes that boolean to `set_random_seed` in the `ResolvedCluster` call on line 667. We could instead have `calc_current_initial_ratio` have an integer argument called `seed` and then include the conversion stated above within the `calc_current_initial_ratio`. However this gives the false impression that the different values of `seed` to this function would have different outcomes when they wouldn't.

My opinion is that it is more clear to the purpose of `calc_current_initial_ratio` and `current_initial_ratio` to have a debug boolean argument instead of a seed integer argument where the different values of the integer leads to the same result. This requires converting the seed integer into a debug boolean before these functions are called. This current commit makes this conversion at the beginning of `_make_comp_dict` on line 782.
